src/ripple/app/tx/impl/OfferStream.cpp

```diff
@@ -222,6 +292,66 @@ TOfferStreamBase<TIn, TOut>::step()
                     << "Removing became unfunded offer " << entry->key();
             }
             offer_ = TOffer<TIn, TOut>{};
+            // See comment at top of loop for how the offer is removed
+            continue;
+        }
+
+        bool const rmSmallIncreasedQOffer = [&] {
+            bool const inIsXRP = isXRP(offer_.issueIn());
+            bool const outIsXRP = isXRP(offer_.issueOut());
+            if (inIsXRP && !outIsXRP)
+            {
+                // Without the `if constexpr`, the
+                // `shouldRmSmallIncreasedQOffer` template will be instantiated
+                // even if it is never used. This can cause compiler errors in
+                // some cases, hence the `if constexpr` guard.
+                if constexpr (!(std::is_same_v<TIn, IOUAmount> ||
+                                std::is_same_v<TOut, XRPAmount>))
+                    return shouldRmSmallIncreasedQOffer<XRPAmount, IOUAmount>();
```

Comment:
There are three questions:
1) Why not `(!std::is_same_v<TIn, IOUAmount> && !std::is_same_v<TOut, XRPAmount>)`
 
Sometimes I write `!(error_condition1 || error_condition2)` instead of `(!error_condition1 && !error_conditions2)`. Think of the huge savings by saving one fewer `!`. More seriously, I don't have a strong preference; I find both about equally readable.

1) Why not `(std::is_same_v<TIn, XRPAmount> && std::is_same_v<TOut, IOUAmount>)`
2) 
Because `TIn` can be `XRPAmount` or `STAmount`, and `TOut` can be `IOUAmount` or `STAmount`.

3) Why is the conditional even needed? Under what circumstances do we enter this codepath...

While the codepath would never be executed, the template would still be instantiated, and we'd get compiler errors. In particular, we'd end up calling things like: `toAmount<IOUAmount>(xrpAmount);` Even though we know this will never be executed, the compiler will still try to compiler that and we'd get a compiler error.The `if constexpr` prevents the template from being instantiated at all.